README.md
Measuring the Accuracy of Password Strength Meters
================================

Password strength meters are an important tool to help users choose more secure passwords.
However, strength meters can only then provide reasonable guidance when they are accurate, i.e., their score correctly reflects password strength.
A strength meter with low accuracy may do more harm than good and guide the user to choose passwords with a high score but actual low security.

The preferred method to measure the accuracy of a strength meter is to compare it to an ideal _reference_, measuring the similarity between the reference and _the meter output_.
In our work, [On the Accuracy of Password Strength Meters](https://www.mobsec.ruhr-uni-bochum.de/forschung/veroeffentlichungen/accuracy-password-strength-meters/), we found the **weighted [Spearman's rank correlation coefficient](https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient)** to be a useful candidate to measure the accuracy of a strength meter compared to the ideal reference.

In this repository you find the necessary code to crawl and evaluate a password strength meter's accuracy.
We hope that this code is helpful for meter developers to improve their implementation.

Project Website
-----------
The original paper, a recording of the talk, the slides, and more information can be found on:

* https://password-meter-comparison.org

User Guide
-----------
The code consists of three parts: `crawler`, `post-processing`, and `evaluation`.
As the crawler uses the [**Selenium framework**](https://www.seleniumhq.org/), we automate a headless Google Chrome/Mozilla Firefox browser to crawl a meter using some predefined list of passwords.
In a post-processing step we use a small **Python** script to prepare the crawled data for the evaluation step in **R**.
In the final step, we calculate the weighted Spearman correlation to estimate the accuracy of the crawled meter.

### Obtaining an Ideal _Reference_
While this repository contains some example passwords from the LinkedIn leak, a more evolved evaluation requires more than one ground truth. The easiest way to obtain such references is by sampling passwords from password leaks, and evaluating them using [CMU's Password Guessability Service (PGS)](https://pgs.ece.cmu.edu/). For more details please refer to [the paper](https://www.mobsec.ruhr-uni-bochum.de/forschung/veroeffentlichungen/accuracy-password-strength-meters/).

### Installation

To keep the guide short, we assume the use of a [Ubuntu-based OS](https://xubuntu.org/release/18-04/).
All `Python` code snippets were tested using **2.7**, all `R` scripts assume version **3.4** or later.

Check out the source code via:

`$ git clone https://github.com/RUB-SysSec/Password-Strength-Meter-Accuracy.git PSMA`

```
.
└── src
    ├── analyze
    │   ├── 01_build_r_file.py
    │   └── 02_corr-comp.r
    ├── crawl
    │   └── 01_zxcvbn
    │       ├── zxcvbn_chrome.py
    │       └── zxcvbn_firefox.py
    ├── datasets
    │   ├── offline
    │   │   ├── 0_linkedin.offline.pw
    │   │   ├── 1_linkedin.offline.strength
    │   │   ├── 2_linkedin.offline.weight
    │   │   └── 3_linkedin.offline.withcount
    │   └── online
    │       ├── 0_linkedin.online.pw
    │       ├── 1_linkedin.online.strength
    │       ├── 2_linkedin.online.weight
    │       └── 3_linkedin.online.withcount
    └── meter
        └── 01_zxcvbn
            ├── index.html
            └── zxcvbn_v4.4.2.js
```
#### Step 0: Preparation

Before we start, we need to install some dependencies, like Python PIP, Selenium, and a WebDriver for your browser.

###### Configuring Python and Installing Selenium
First we install Python [Package Installer](https://pip.pypa.io/en/stable/) (PIP) and the Python [virtual environment](https://docs.python.org/2.7/glossary.html#term-virtual-environment) runtime environment.

`$ sudo apt-get install python-pip virtualenv`

We start by creating a new Python virtual environment that we just use for this project.

`$ virtualenv -p /usr/bin/python2.7 venv`

Next, we activate the Python virtual environment.

`$ source venv/bin/activate`

Now, we install selenium.

`(venv) $ pip install selenium`

###### Installing the WebDriver

To allow Selenium to communicate and automate your browser, we need to install your web browser's driver.

* The WebDriver for Google Chrome can be found [here](http://chromedriver.chromium.org/downloads).
* The WebDriver for Mozilla Firefox can be found [here](https://github.com/mozilla/geckodriver/releases).

Next, download and install the driver you need depending on the browser you like to use!

Below is an example for Google Chrome:

```
$ wget https://chromedriver.storage.googleapis.com/<release>/chromedriver_linux64.zip
$ unzip chromedriver_linux64.zip
$ chmod +x chromedriver
$ sudo mv chromedriver /usr/local/bin/
```

###### Installing R

Install the [R open-source programming language](https://www.digitalocean.com/community/tutorials/how-to-install-r-on-ubuntu-18-04-quickstart) on your Ubuntu machine.

Optional: Install [RStudio Desktop](https://www.rstudio.com/products/rstudio/download/#download) (Open Source License) for more convenience.

#### Step 1: Crawling the Meter
In a first step, we need to get the estimates from the strength meter for a given set of passwords.
For most web-based password strength meters we need the Selenium framework and a web browser (and its WebDriver!) to obtain such estimates.
Our tutorial includes an example based on the [zxcvbn strength meter](https://github.com/dropbox/zxcvbn).

While we try to explain everything in detail, we skip the part on how to install a web browser on your system, just use [Google Chrome](https://www.google.com/chrome/) or [Mozilla Firefox](https://www.mozilla.org/en-US/firefox/new/).

Note: If your meter is not a web-based strength meter, you need write some custom function that will output strength estimates for a given list of passwords found in the `datasets` folder.

Navigate to `src/meter/01_zxcvbn/` and open `index.html` with your browser. Copy the path that is displayed in the URL bar, likely similar to:

`file:///home/<username>/PSMA/src/meter/01_zxcvbn/index.html`

Now edit `src/crawl/01_zxcvbn/zxcvbn_chrome.py` or `src/crawl/01_zxcvbn/zxcvbn_firefox.py` and change the path according to your environment.

```
...
    driver.get('file:///home/<username>/PSMA/src/meter/01_zxcvbn/index.html')
...
```

Save your edits!

Next, we make sure that the Python virtual environment is activated, and we change the current directory to `src/crawl/01_zxcvbn/`.

```
$ cd src/crawl/01_zxcvbn/
$ source ~/venv/bin/activate
(venv) $ python zxcvbn_chrome.py ../../datasets/online/0_linkedin.online.pw
```

![Alt text](/docs/screenshots/zxcvbn.gif?raw=true "zxcvbn Crawling in Action")

#### Step 2: Post-Processing

First, navigate to `src/analyze/` and edit `01_build_r_file.py` to your needs.

On the Terminal run:

```
$ cd src/analyze/
$ python 01_build_r_file.py
```

It will produce 2 files:
* `result_online.csv`
* `result_offline.csv`

Contents of `result_online.csv`:
```
count  strength  weight    zxcvbn_guess_number password zxcvbn_score
30.0  -1044164.0 1044164.0 2.0                 123456   0.0
 6.0  -193001.0  193001.0  22802.0             linkedin 1.0
 4.0  -176120.0  176120.0  3.0                 password 0.0
 4.0  -78720.0   78720.0   9.0                 111111   0.0
...
```

#### Step 3: Evaluation using R

Start R, e.g., by running RStudio or `R`:

`> install.packages("wCorr")`

This installs the support for Weighted Correlations in R.

In R run `02_corr-comp.r`, e.g., by running:

`$ Rscript 02_corr-comp.r`

The output for `result_online.csv`:
```
zxcvbn_guess_number 0.738
zxcvbn_score        0.382
```

Now edit `02_corr-comp.r` and change the input file to `"result_offline.csv"`.

The output for `result_offline.csv`:
```
zxcvbn_guess_number 0.896 	
zxcvbn_score        0.689
```

FAQ
---

* The crawling doesn't work!
Make sure you have activated the virtual environment. Check for the presence of the `(venv)` in front of your prompt.
Make sure you have the latest WebDriver installed on your system.

License
-------

Our code in the **Password-Strength-Meter-Accuracy** repository is licensed under the MIT license.
Refer to [docs/LICENSE](docs/LICENSE) for more information.

### Third-Party Libraries
* **zxcvbn** is a password strength meter developed by Daniel Wheeler and Dropbox, Inc. and is using the MIT license.
The license can be found [here](https://github.com/dropbox/zxcvbn/blob/master/LICENSE.txt).
* **jQuery** is a JavaScript library developed by the JS Foundation and is using the MIT license.
The license can be found [here](https://jquery.org/license/).
* **wCorr** is an R package developed by Ahmad Emad and Paul Bailey and is using the GPL-2 license.
The license can be found [here](https://cran.r-project.org/web/licenses/GPL-2).
* **Selenium** is a browser automation framework developed by ThoughtWorks and is using the Apache 2.0 license.
The license can be found [here](https://www.apache.org/licenses/LICENSE-2.0).

Contact
-------
Visit our [website](https://www.mobsec.rub.de) and follow us on [Twitter](https://twitter.com/hgi_bochum).
If you are interested in passwords, consider to contribute and to attend the [International Conference on Passwords (PASSWORDS)](https://passwordscon.org).


docs/CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Planned
- Add more password strength meters

## [0.1.0] - 2019-04-24
### Added
- Initial version of the source code


src/analyze/02_corr-comp.r
rm(list = ls()) # clean workspace
library(wCorr)
# read data
merged <- read.delim("result_online.csv", header = TRUE, quote = "")
# config used data and methods here
LIST_DATA   = c("zxcvbn_guess_number", "zxcvbn_score")
LIST_METHOD = c("wspearman")
REF    = merged$strength # used as refernce
WEIGHT = merged$weight   # used as weight
###########################
results = matrix(numeric(0),
                 nrow = length(LIST_DATA),
                 ncol = length(LIST_METHOD))
# main loop
row_index = 0
for (COL in LIST_DATA) {
  row_index = row_index + 1
  cat(COL, "\t")
  SAM = merged[, COL]
  col_index = 0
  for (METHOD in LIST_METHOD) {
    col_index = col_index + 1
    if (METHOD == "wspearman") {
      C = weightedCorr(REF, SAM, method = "spearman", weights = WEIGHT)
    } else{
      C = -1
    }
    cat(round(C, digits = 3), "\t") # output to stdout
  }
  cat("\n")
}
# -1.0 means strong negative correlation; meter works but 'strong' passwords are in fact 'weak' and the other way round
#  0.0 means no correlation; meter is randomly guessing, and not estimating password strength
#  1.0 means strong positive correlation; meter works perfectly

src/meter/01_zxcvbn/eval.js
function checkStrength(password) {
    'use strict';
    var user_inputs = []; // could be the username, email address, service name
    var result = zxcvbn(password, user_inputs);
    document.getElementById("guess_number").innerHTML = result.guesses;
    document.getElementById("score").innerHTML = result.score;
}

$(document).ready(function () {
    'use strict';
    // your function
    var my_func = function (event) {
        event.preventDefault(); // stop default submit action of HTML forms
        // reset the form to be fresh
        document.getElementById("guess_number").innerHTML = "";
        document.getElementById("score").innerHTML = "";
        // get password and check strength / blacklist
        var password = document.getElementById("password").value;
        checkStrength(password);
    };
    // your form
    var form = document.getElementById("myForm");
    // attach event listener
    form.addEventListener("submit", my_func, true);
});


src/crawl/01_zxcvbn/zxcvbn_chrome.py
import time, sys, random, re, base64
from selenium import webdriver
# Identify DOM elements By.ID, By.XPATH, ...
from selenium.webdriver.common.by import By
# Wait for DOM elements to appear/be visible
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException

# Most common user agents: https://techblog.willshouse.com/2012/01/03/most-common-user-agents/
def init_driver(defaultUseragent='Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/73.0.3683.86 Safari/537.36'):
    # Chrome options
    options = webdriver.ChromeOptions()
    #options.add_argument('headless') # Headless, activate to hide browser window from interrupting your normal work
    options.add_argument('window-size=1920x1080')
    options.add_experimental_option('prefs', {'intl.accept_languages': 'en,en_US'})
    #options.add_argument('proxy-server=socks5://127.0.0.1:1080')
    options.add_argument('user-agent=' + defaultUseragent)
    driver = webdriver.Chrome(chrome_options=options)
    return driver

def main():
    driver = init_driver()

    driver.get('file:///home/<username>/PSMA/src/meter/01_zxcvbn/index.html')
    delay = 5 # seconds

    try:
        pass_field = WebDriverWait(driver, delay).until(EC.element_to_be_clickable((By.ID, 'password'))) #presence_of_element_located
    except TimeoutException:
        print("Error: Element with id='password' was not found within {} seconds.").format(delay)

    try:
        submit_button = WebDriverWait(driver, delay).until(EC.element_to_be_clickable((By.ID, 'submit'))) #presence_of_element_located
    except TimeoutException:
        print("Error: Element with id='submit' was not found within {} seconds.").format(delay)

    fog = open("./"+str(sys.argv[1].split('/')[-1])+"_guess_number_result.txt", "w")
    fog.write("{}\n".format("zxcvbn_guess_number_linkedin"))
    fos = open("./"+str(sys.argv[1].split('/')[-1])+"_score_result.txt", "w")
    fos.write("{}\n".format("zxcvbn_score_linkedin"))
    with open(str(sys.argv[1]),'rU') as inputfile:
        inputfile.readline() # skip header
        for line in inputfile:
            line = line.rstrip('\r\n')
            # Send new password to evaluate
            pass_field.send_keys(line)
            submit_button.click()
            #time.sleep(0.3) # Some Ajax meters require some processing time
            # Search for result on page
            guess_number = "-1.0"
            try:
                guess_number_field = WebDriverWait(driver, delay).until(EC.element_to_be_clickable((By.ID, 'guess_number'))) #presence_of_element_located
            except TimeoutException:
                print("Error: Element with id='guess_number' was not found within {} seconds.").format(delay)
            guess_number = guess_number_field.get_attribute("innerHTML")
            score = "-1.0"
            try:
                score_field = WebDriverWait(driver, delay).until(EC.element_to_be_clickable((By.ID, 'score'))) #presence_of_element_located
            except TimeoutException:
                print("Error: Element with id='score' was not found within {} seconds.").format(delay)
            score = score_field.get_attribute("innerHTML")
            # Print result to stdout
            print("{}\t{}\t{}".format(score, guess_number, line))
            # Write result to file
            fog.write("{}\n".format(guess_number))
            fos.write("{}\n".format(score))
            pass_field.clear() # reset
            fog.flush()
            fos.flush()

    fog.close()
    fos.close()
    driver.quit()

if __name__ == "__main__":
    main()


src/crawl/01_zxcvbn/zxcvbn_firefox.py
import time, sys, random, re, base64
from selenium import webdriver
# Identify DOM elements By.ID, By.XPATH, ...
from selenium.webdriver.common.by import By
# Wait for DOM elements to appear/be visible
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.common.exceptions import TimeoutException

# Most common user agents: https://techblog.willshouse.com/2012/01/03/most-common-user-agents/
def init_driver(defaultUseragent='Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:66.0) Gecko/20100101 Firefox/66.0'):
    # Firefox options
    from selenium.webdriver.firefox.firefox_binary import FirefoxBinary
    from selenium.webdriver.firefox.options import Options
    binary = FirefoxBinary('/usr/bin/firefox') # Change to your needs
    options = Options()
    #options.headless = True  # Headless, activate to hide browser window from interrupting your normal work
    profile = webdriver.FirefoxProfile()
    profile.set_preference("intl.accept_languages", 'en,en_US')
    #profile.set_preference("network.proxy.type", 1)
    #profile.set_preference("network.proxy.socks", "127.0.0.1")
    #profile.set_preference("network.proxy.socks_port", 1080)
    #profile.set_preference("network.proxy.socks_version", 5)
    profile.set_preference("general.useragent.override", defaultUseragent)
    profile.update_preferences()
    driver = webdriver.Firefox(options=options, firefox_binary=binary, firefox_profile=profile)
    driver.set_window_size(1920, 1080)
    return driver

def main():
    driver = init_driver()

    driver.get('file:///home/<username>/PSMA/src/meter/01_zxcvbn/index.html')
    delay = 5 # seconds

    try:
        pass_field = WebDriverWait(driver, delay).until(EC.element_to_be_clickable((By.ID, 'password'))) #presence_of_element_located
    except TimeoutException:
        print("Error: Element with id='password' was not found within {} seconds.").format(delay)

    try:
        submit_button = WebDriverWait(driver, delay).until(EC.element_to_be_clickable((By.ID, 'submit'))) #presence_of_element_located
    except TimeoutException:
        print("Error: Element with id='submit' was not found within {} seconds.").format(delay)

    fog = open("./"+str(sys.argv[1].split('/')[-1])+"_guess_number_result.txt", "w")
    fog.write("{}\n".format("zxcvbn_guess_number_linkedin"))
    fos = open("./"+str(sys.argv[1].split('/')[-1])+"_score_result.txt", "w")
    fos.write("{}\n".format("zxcvbn_score_linkedin"))
    with open(str(sys.argv[1]),'rU') as inputfile:
        inputfile.readline() # skip header
        for line in inputfile:
            line = line.rstrip('\r\n')
            # Send new password to evaluate
            pass_field.send_keys(line)
            submit_button.click()
            #time.sleep(0.3) # Some Ajax meters require some processing time
            # Search for result on page
            guess_number = "-1.0"
            try:
                guess_number_field = WebDriverWait(driver, delay).until(EC.element_to_be_clickable((By.ID, 'guess_number'))) #presence_of_element_located
            except TimeoutException:
                print("Error: Element with id='guess_number' was not found within {} seconds.").format(delay)
            guess_number = guess_number_field.get_attribute("innerHTML")
            score = "-1.0"
            try:
                score_field = WebDriverWait(driver, delay).until(EC.element_to_be_clickable((By.ID, 'score'))) #presence_of_element_located
            except TimeoutException:
                print("Error: Element with id='score' was not found within {} seconds.").format(delay)
            score = score_field.get_attribute("innerHTML")
            # Print result to stdout
            print("{}\t{}\t{}".format(score, guess_number, line))
            # Write result to file
            fog.write("{}\n".format(guess_number))
            fos.write("{}\n".format(score))
            pass_field.clear() # reset
            fog.flush()
            fos.flush()

    fog.close()
    fos.close()
    driver.quit()

if __name__ == "__main__":
    main()


src/analyze/01_build_r_file.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

'''
:author: Maximilian Golla
:contact: maximilian.golla@rub.de
:version: 0.0.2, 2019-04-24
'''

import sys, re
from collections import OrderedDict

def read_file(filename):
    data = []
    with open(filename, 'r') as inputfile:
        inputfile.readline() # skip header
        for line in inputfile:
            try:
                line = line.rstrip('\r\n')
                data.append(line)
            except Exception as e:
                sys.stderr.write("Error: {} - {}\n".format(e, line))
    return data

def _prepare_dict(strength, weight, withcount):
    pw_re = re.compile('^\s*[0-9]*\s')
    occ_re = re.compile('^\s*[0-9]*')

    output = OrderedDict()
    for i in xrange(0, len(strength)):
        p = withcount[i].replace(pw_re.findall(withcount[i])[0], '')
        s = float(strength[i])
        w = float(weight[i])
        c = float(occ_re.findall(withcount[i])[0].replace(' ', ''))
        output[i] = {'password': p, 'strength': s, 'weight': w, 'count': c}
    return output

def _add_meter(output, meter_name, meter_result_file):
    meter_data = read_file(meter_result_file)
    i = 0
    for entry in output:
        output[entry][meter_name] = float(meter_data[i])
        i += 1
    return output

def build_online():
    strength = read_file("../datasets/online/1_linkedin.online.strength")
    weight = read_file("../datasets/online/2_linkedin.online.weight")
    withcount = read_file("../datasets/online/3_linkedin.online.withcount")
    output = _prepare_dict(strength, weight, withcount)
    output = _add_meter(output, "zxcvbn_guess_number", "../crawl/01_zxcvbn/0_linkedin.online.pw_guess_number_result.txt")
    output = _add_meter(output, "zxcvbn_score", "../crawl/01_zxcvbn/0_linkedin.online.pw_score_result.txt")
    return output

def build_offline():
    strength = read_file("../datasets/offline/1_linkedin.offline.strength")
    weight = read_file("../datasets/offline/2_linkedin.offline.weight")
    withcount = read_file("../datasets/offline/3_linkedin.offline.withcount")
    output = _prepare_dict(strength, weight, withcount)
    output = _add_meter(output, "zxcvbn_guess_number", "../crawl/01_zxcvbn/0_linkedin.offline.pw_guess_number_result.txt")
    output = _add_meter(output, "zxcvbn_score", "../crawl/01_zxcvbn/0_linkedin.offline.pw_score_result.txt")
    return output

def write_result(filename, result):
    with open(filename, 'w') as outputfile:
        # Write header
        header = []
        for title in result.itervalues().next():
            header.append(title)
        outputfile.write("{}\n".format("\t".join(header)))
        # Write data
        line_number = 1
        for entry in result:
            outputfile.write("{}\n".format("\t".join([str(i) for i in result[entry].values()])))
            line_number += 1

def main():
    result = build_online()
    write_result("result_online.csv", result)
    result = build_offline()
    write_result("result_offline.csv", result)

if __name__ == '__main__':
    main()
